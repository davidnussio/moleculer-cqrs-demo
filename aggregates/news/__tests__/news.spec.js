const { CQRSFixture } = require("moleculer-cqrs");

const aggregate = require("..");

const {
  commands: { createNews, upvoteNews },
  events: { NewsCreatedEvent, NewsDeletedEvent },
} = aggregate;

jest
  .spyOn(global.Date, "now")
  .mockImplementation(() => new Date("2019-05-14T11:01:58.135Z").valueOf());

describe("Testing aggregate commands in isolation", () => {
  test("should commands with empty payload throw error", () => {
    expect(() => createNews({}, {})).toThrow("Aggregate validation error");
  });

  test("should createNews command return an NewsCreatedEvent", () => {
    expect(
      createNews(
        {},
        {
          payload: {
            title: "Title news",
            userId: "user-id-1",
            text:
              "Asperiores nam tempora qui et provident temporibus illo et fugit.",
          },
        }
      )
    ).toMatchSnapshot();
  });
});

describe("Testing news aggregate with cqrs fixture", () => {
  let fixture;

  beforeEach(() => {
    fixture = new CQRSFixture(aggregate);
  });

  test("should call raw command", () => {
    fixture
      .givenEvents([])
      .when({
        aggregateId: "12345",
        aggregateName: "news",
        type: "createNews",
        payload: {
          title: "Title news... title...",
          text: "Text text... text... text...",
          userId: "user-id-1abc2def3",
        },
      })
      .expectEvents(
        NewsCreatedEvent(
          "Title news... title...",
          "Text text... text... text...",
          "",
          [],
          "user-id-1abc2def3",
          Date.now()
        )
      );
  });

  test("should command function", () => {
    fixture
      .givenEvents()
      .when(createNews, {
        title: "Title news... title...",
        text: "Text text... text... text...",
        userId: "user-id-1abc2def3",
      })
      .expectEvents(
        NewsCreatedEvent(
          "Title news... title...",
          "Text text... text... text...",
          "",
          [],
          "user-id-1abc2def3",
          Date.now()
        )
      );
  });

  test("should reject upvode command when news already deleted", () => {
    fixture
      .givenEvents([
        NewsCreatedEvent(
          "Title news... title...",
          "Text text... text... text...",
          "",
          [],
          "user-id-1abc2def3",
          Date.now()
        ),
        NewsDeletedEvent(Date.now()),
      ])
      .whenThrow(upvoteNews, { userId: "user-1234" })
      .toThrow("Aggregate is already deleted");
  });
});
