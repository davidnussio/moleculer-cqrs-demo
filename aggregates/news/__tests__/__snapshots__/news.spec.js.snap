// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing aggregate commands in isolation should createNews command return an NewsCreatedEvent 1`] = `
Object {
  "payload": Object {
    "createdAt": 1557831718135,
    "createdBy": "user-id-1",
    "link": "",
    "text": "Asperiores nam tempora qui et provident temporibus illo et fugit.",
    "title": "Title news",
    "votes": Array [],
  },
  "type": "aggregate.news.created",
}
`;
